addons/l10n_tw_edi_ecpay/models/account_move.py

```diff
@@ -0,0 +1,786 @@
+        # EXTENDS 'account'
+        invoices_to_reset = self.filtered(
+            lambda i: (i.state == "cancel" and i.l10n_tw_edi_state == "invalid")
+        )
+        res = super().button_draft()
+
+        invoices_to_reset.write({
+            "l10n_tw_edi_related_number": False,
+            "l10n_tw_edi_state": False,
+            "l10n_tw_edi_ecpay_invoice_id": False,
+            "l10n_tw_edi_invoice_create_date": False,
+            "l10n_tw_edi_invalidate_reason": False,
+        })
+        invoices_to_reset.l10n_tw_edi_file_id.unlink()
+        return res
+
+    @api.depends("l10n_tw_edi_love_code", "l10n_tw_edi_carrier_type", "partner_id")
+    def _compute_is_print(self):
+        for move in self:
+            if move.l10n_tw_edi_love_code or (move.partner_id.vat and move.l10n_tw_edi_carrier_type in [1, 2]):
+                move.l10n_tw_edi_is_print = False
+
+    @api.depends("l10n_tw_edi_is_print", "l10n_tw_edi_carrier_type", "partner_id")
+    def _compute_love_code(self):
+        for move in self:
+            if move.l10n_tw_edi_is_print or move.l10n_tw_edi_carrier_type or move.partner_id.vat:
+                move.l10n_tw_edi_love_code = False
+
+    @api.depends("l10n_tw_edi_is_print", "l10n_tw_edi_love_code")
+    def _compute_carrier_info(self):
+        for move in self:
+            if move.l10n_tw_edi_is_print or move.l10n_tw_edi_love_code:
+                move.l10n_tw_edi_carrier_type = False
+                move.l10n_tw_edi_carrier_number = False
+                move.l10n_tw_edi_carrier_number_2 = False
+
+    @api.depends("invoice_line_ids.tax_ids.l10n_tw_edi_tax_type")
+    def _compute_l10n_tw_edi_tax_info(self):
+        for move in self:
+            tax_type, special_tax_type, is_zero_tax_rate = move._l10n_tw_edi_compute_tax_type_from_invoice_lines()
+            if tax_type == "3":
+                move.l10n_tw_edi_invoice_type = "07" if not special_tax_type else "08"
+            else:
+                move.l10n_tw_edi_invoice_type = "07" if tax_type != "4" else "08"
+            move.l10n_tw_edi_is_zero_tax_rate = is_zero_tax_rate if move.invoice_line_ids.tax_ids else False
+
+    @api.depends("partner_id.parent_id", "partner_id.company_type")
+    def _compute_l10n_tw_edi_is_b2b(self):
+        for rec in self:
+            rec.l10n_tw_edi_is_b2b = (rec.partner_id.parent_id and rec.partner_id.parent_id.company_type == "company") or rec.partner_id.company_type == "company"
+
+    # ----------------
+    # Business methods
+    # ----------------
+    def _get_mail_thread_data_attachments(self):
+        res = super()._get_mail_thread_data_attachments()
+        return res | self.l10n_tw_edi_file_id
+
+    # API methods
+    def _l10n_tw_edi_rounding(self, num):
+        """
+        Round the number to two decimal places using ROUND_HALF_UP (3.115 -> 3.12, 3.114 -> 3.11)
+        """
+        return float(Decimal(str(num)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP))
```

Comment:
Use float_round from float_utils. it default to half up. And then this method is no longer useful.